code/modules/mob/living/simple_animal/distortion/star/TanyaBoss.dm

```diff
@@ -0,0 +1,214 @@
+	if(!can_act)
+		return FALSE
+
+	if(client)
+		switch(chosen_attack)
+			if(1)
+				OnRush(target)
+		return
+
+	if(onrush_cooldown <= world.time)
+		OnRush(target)
+		return
+
+
+	return
+
+/mob/living/simple_animal/hostile/distortion/Tanya/death()
+	animate(src, alpha = 0, time = (10 SECONDS))
+	QDEL_IN(src, (10 SECONDS))
+	return ..()
+
+// Copy and paste Red Mist
+/mob/living/simple_animal/hostile/distortion/Tanya/proc/OnRush(mob/living/target, iteration = 0)
+	if(!can_act && !iteration)
+		return
+
+	if(!iteration && onrush_cooldown > world.time)
+		return
+
+	if(iteration >= onrush_max_iterations)
+		can_act = TRUE
+		return
+
+	can_act = FALSE
+	onrush_cooldown = world.time + onrush_cooldown_time
+	face_atom(target)
+	if(!iteration)
+		var/obj/effect/temp_visual/decoy/D = new /obj/effect/temp_visual/decoy(get_turf(src), src)
+		animate(D, alpha = 0, transform = matrix()*1.5, time = 3)
+		visible_message(span_danger("[src] prepares to beat everyone to a pulp!"))
+		onrush_hit = list()
+		SLEEP_CHECK_DEATH(5)
+
+	if(!istype(target))
+		can_act = TRUE
+		return
+
+	var/turf/target_turf = get_step(target, get_dir(target, src))
+	var/list/line_list = getline(src, target_turf)
+	for(var/i = 1 to length(line_list))
+		var/turf/TT = line_list[i]
+		var/obj/effect/temp_visual/decoy/D = new (TT, src)
+		D.alpha = min(150 + i*15, 255)
+		animate(D, alpha = 0, time = 4 + i*2)
+
+	forceMove(target_turf)
+	// PUNCH!!!!
+	playsound(src, pick(onrush_sounds), rand(50, 100), TRUE, 7)
+	to_chat(target, span_userdanger("[src] punches you hard!"))
+	if(!onrush_double_hit)
+		onrush_hit |= target
+	var/turf/thrownat = get_ranged_target_turf_direct(src, target, 15, rand(-30, 30))
+	target.throw_at(thrownat, 8, 2, src, TRUE, force = MOVE_FORCE_OVERPOWERING, gentle = FALSE)
+	target.deal_damage(onrush_damage, RED_DAMAGE)
+	new /obj/effect/temp_visual/smash_effect(get_turf(target))
+	shake_camera(target, 2, 5)
+
+	if(iteration >= onrush_max_iterations)
+		can_act = TRUE
+		onrush_hit = list()
+		return
+
+	// Pick new target
+	var/list/valid_targets = list()
+	for(var/mob/living/L in view(9, src))
+		if(L.stat == DEAD)
+			continue
+		if(faction_check_mob(L))
+			continue
+		if((L in onrush_hit) && !onrush_double_hit)
+			continue
+		if(L == target)
+			continue
+		valid_targets += L
+
+	if(!LAZYLEN(valid_targets))
+		can_act = TRUE
+		onrush_hit = list()
+		return
+
+	var/mob/living/L = pick(valid_targets)
+	var/obj/effect/temp_visual/decoy/D = new /obj/effect/temp_visual/decoy(get_turf(L), L)
+	animate(D, alpha = 0, transform = matrix()*1.2, time = 3)
+	face_atom(L)
+	addtimer(CALLBACK(src, PROC_REF(OnRush), L, iteration + 1), rand(onrush_min_delay, onrush_max_delay))
+	return
+
+/mob/living/simple_animal/hostile/distortion/Tanya/death()
+	for(var/V in fire_list)
+		fire_list-=V
+		qdel(V)
+	..()
+
+/mob/living/simple_animal/hostile/distortion/Tanya/Move()
+	..()
+	for(var/obj/structure/spreading/Tanya_sand/Y in get_turf(src))
```

Comment:
You can instead use locate(var/obj/structure/spreading/Tanya_sand) in get_turf(src) to see if sand is there.